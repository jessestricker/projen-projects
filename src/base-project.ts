import { Project, ProjectOptions, javascript, typescript } from "projen";

export interface BaseProjectOptions extends ProjectOptions {
  /**
   * The Node package manager used to install dependencies and run scripts.
   *
   * @default javascript.NodePackageManager.NPM
   */
  readonly packageManager?: javascript.NodePackageManager;

  /**
   * The NPM dependencies used for development.
   */
  readonly devDeps?: string[];
}

export class BaseProject extends Project {
  constructor(options: BaseProjectOptions) {
    if (options.projenrcJson || options.projenrcJsonOptions) {
      throw new Error(".projenrc.json not supported");
    }

    const packageManager =
      options.packageManager ?? javascript.NodePackageManager.NPM;
    const packageRunnerCommand = getPackageRunnerCommand(packageManager);

    super({
      ...options,
      projenCommand: `${packageRunnerCommand} projen`,
    });

    // node package
    const nodePackage = new javascript.NodePackage(this, {
      devDeps: options.devDeps,
      entrypoint: "",
      packageManager: packageManager,
      licensed: false,
    });
    nodePackage.addField("private", true);
    nodePackage.addField("name", undefined);
    nodePackage.addField("publishConfig", undefined);
    nodePackage.addField("version", undefined);

    nodePackage.addDevDeps("projen", "constructs@^10.0.0");

    // projenrc file
    const projenrc = new typescript.ProjenrcTs(this);
    nodePackage.addDevDeps("ts-node");
    this.defaultTask?.reset(
      `${packageRunnerCommand} ts-node --project ${projenrc.tsconfig.fileName} ${projenrc.filePath}`,
    );
  }
}

function getPackageRunnerCommand(
  packageManager: javascript.NodePackageManager,
) {
  switch (packageManager) {
    case javascript.NodePackageManager.YARN_BERRY:
      return "yarn dlx";
    case javascript.NodePackageManager.NPM:
      return "npx";
    case javascript.NodePackageManager.PNPM:
      return "pnpx";
    case javascript.NodePackageManager.BUN:
      return "bunx";
    default:
      throw new Error(`unsupported package manager "${packageManager}"`);
  }
}
