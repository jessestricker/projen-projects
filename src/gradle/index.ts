export * from "./gradle";
export * from "./gradle-project";
