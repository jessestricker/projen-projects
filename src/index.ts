export * from "./base-project";
export * from "./gradle";
