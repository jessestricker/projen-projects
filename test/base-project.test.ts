import { Testing } from "projen";
import { BaseProject } from "../src";

describe("BaseProject", () => {
  test("matches snapshot", () => {
    const project = new BaseProject({
      name: "base-project",
    });

    const snapshot = Testing.synth(project);

    expect(snapshot).toMatchSnapshot();
  });
});
