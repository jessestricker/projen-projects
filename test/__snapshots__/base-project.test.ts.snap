// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseProject matches snapshot 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/package.json
!/tsconfig.projen.json
!/.projenrc.ts
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "dependencies": [
      {
        "name": "constructs",
        "type": "build",
        "version": "^10.0.0",
      },
      {
        "name": "projen",
        "type": "build",
      },
      {
        "name": "ts-node",
        "type": "build",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "tsconfig.projen.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
        "steps": [
          {
            "exec": "npx ts-node --project tsconfig.projen.json .projenrc.ts",
          },
        ],
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install project dependencies and update lockfile (non-frozen)",
        "name": "install",
        "steps": [
          {
            "exec": "npm install",
          },
        ],
      },
      "install:ci": {
        "description": "Install project dependencies using frozen lockfile",
        "name": "install:ci",
        "steps": [
          {
            "exec": "npm ci",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "package.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "devDependencies": {
      "constructs": "^10.0.0",
      "projen": "*",
      "ts-node": "*",
    },
    "license": "UNLICENSED",
    "private": true,
    "scripts": {
      "build": "npx projen build",
      "compile": "npx projen compile",
      "default": "npx projen default",
      "eject": "npx projen eject",
      "package": "npx projen package",
      "post-compile": "npx projen post-compile",
      "pre-compile": "npx projen pre-compile",
      "test": "npx projen test",
    },
  },
  "tsconfig.projen.json": {
    "compilerOptions": {},
    "exclude": [
      "node_modules",
    ],
    "include": [
      "**/*.ts",
      ".projenrc.ts",
      "projenrc/**/*.ts",
    ],
  },
}
`;
